### DISCUSSION TITLE
Documenting the OCPP Integration

### DISCUSSION THREAD

**Discussion by GrantKeymer** [lbbrhzn/ocpp#376] [Ideas]
I've noticed since the recent changes to a separate **sensor.charger.status.connector** that the original one **sensor.charger_status** always says 'Available' no matter what.  It had me fooled for a while because everything relying on the state of that sensor was broken.  My fancy dashboard didn't work and even the automation written by @lbbrhzn didn't work 😕 

Then I saw the pull request where the new sensor was mentioned and it seems to work fine.

Should the original sensor be removed to avoid confusion?

**Comment by lbbrhzn** [comment]
Hi Grant, sorry to break your automations! I probably should have documented this more clearly in the release notes, my apologies. It seems there is a need to document the integration in more detail, for instance in the project's wiki.
Maybe you'd like to volunteer?

According to the OCPP standard there really is more than a single status: one for the main controller of the charger, and one for each connector. Yes, that means there can even be more than two!

Previously there was only a single status entity, resulting in a continuously toggling between several status values for some systems.

To solve that issue I decided to map everything on just two status entities. One for the main status (connector_id=0) and one for the connector statuses (connector_id>0). The state value of the connector status entity reflects connector 1, other connectors are stored in the extra attributes. 

There's a similar issue with the power / energy / current / voltage values: there can be different ones, depending on the connector and the transaction.

I'm still trying to figure out how to present this in the UI in a simple way, without creating a plethora of entities. Any suggestions from you side?

**Reply by GrantKeymer** [reply]
> Hi Grant, sorry to break your automations! I probably should have documented this more clearly in the release notes, my apologies. It seems there is a need to document the integration in more detail, for instance in the project's wiki. Maybe you'd like to volunteer?
> 
Thanks for asking.  Yes, I'd be happy to take that project on.

> According to the OCPP standard there really is more than a single status: one for the main controller of the charger, and one for each connector. Yes, that means there can even be more than two!
> 
> Previously there was only a single status entity, resulting in a continuously toggling between several status values for some systems.
> 
> To solve that issue I decided to map everything on just two status entities. One for the main status (connector_id=0) and one for the connector statuses (connector_id>0). The state value of the connector status entity reflects connector 1, other connectors are stored in the extra attributes.
> 
> There's a similar issue with the power / energy / current / voltage values: there can be different ones, depending on the connector and the transaction.
> 
> I'm still trying to figure out how to present this in the UI in a simple way, without creating a plethora of entities. Any suggestions from you side?
>
I'll give that some thought and get back to you.  I guess some 3-phase chargers have separate meter values for each phase as well?

In coming up with a solution, we need to design a structure which is flexible enough to accommodate single phase, 3-phase, maybe even split-phase as they have in the US.  And then to allow for different connectors as you say.  Hmmm, time to put the thinking cap on 🙂

**Reply by GrantKeymer** [reply]
> I'm still trying to figure out how to present this in the UI in a simple way, without creating a plethora of entities. Any suggestions from you side?
>
After giving this quite a bit of thought, I can see now that the `extra_state_attributes` are stored as a Python `dict` class, with a set of Key/Value pairs.  That seems to be a great way to store all the data from extra connectors or phases, as the case may be.

I can also see the sense in avoiding a plethora of entities.  There are plenty now but if we needed to add another 10 or so, it would get pretty unwieldy.  So please explain this...

**Is it possible to add a Sensor Entity to a Lovelace Card which contains a 'dict' object, so the appropriate value may be selected during card editing and displayed?**

If the answer is yes, it becomes a relatively simple matter of working out a logical naming system, which I can definitely help with.

**Reply by lbbrhzn** [reply]
The attributes can be retireved trough a template function: `device_attr(device_or_entity_id, attr_name)`
The way to get this in a card is to define a template sensor first. See
https://www.home-assistant.io/integrations/template/

There are custom card that allow you to do templating directly:
https://github.com/gadgetchnnel/lovelace-card-templater

**Reply by drc38** [reply]
Might be worth looking at ApexCharts too https://community.home-assistant.io/t/apexcharts-card-a-highly-customizable-graph-card/272877

**Reply by GrantKeymer** [reply]
Some excellent tips there guys, thanks!  As usual, someone has already spent the time and created just what we need to solve the problem of displaying the measurands stored in extra_attr.

I like the approach you have taken of keeping things very simple for the most basic use case of a single connector.  It avoids needless complexity for home users.  However, for parking buildings where each charger can have 2 or more connectors, we need to come up with a suitable naming convention to identify the additional measurands.

How about I propose a solution and see what you think?

Do you know of any chargers which provide separate metering for each phase?  Or would that be unnecessary complexity at this point?

**Reply by lbbrhzn** [reply]
We support phase info already, and store it in the extra attributes. If only phase info is passed, we take the average and use that as the state value.

**Reply by GrantKeymer** [reply]
Here's my suggestion for how to improve the naming convention regarding various controls and sensors...

**Controls relating to the entire charger**
```
button.charger_reset
switch.charger_availability
```
**Controls relating to a particular connector where there is only one**
```
switch.charger_charge_control
number.charger_maximum_current
button.charger_unlock
```
**Controls relating to a particular connector where there is more than one**
```
switch.charger_connector2_charge_control (or connector3 etc)
number.charger_connector2_maximum_current (or connector3 etc)
button.charger_connector2_unlock (or connector3 etc)
```
**Sensors relating to the entire charger**
```
charger_frequency
charger_voltage
charger_energy_import_register
charger_energy_meter_start
charger_connectors
charger_status
charger_latency_ping
charger_latency_pong
charger_heartbeat
charger_reconnects
charger_id
charger_id_tag
charger_error_code
charger_model
charger_features
charger_serial
charger_status_firmware
charger_timestamp_config_response
charger_timestamp_data_response
charger_vendor
charger_version_firmware
```
**Sensors relating to a particular connector where there is only one**
```
charger_power_active_import
charger_power_offered
charger_time_session
charger_transaction_id
charger_current_import
charger_current_offered
charger_energy_session
charger_connector_status (** CHANGED **)
charger_stop_reason
charger_connector_error_code (** CHANGED **)
```
**Sensors relating to a particular connector where there is more than one**
```
charger_connector2_power_active_import (or connector3 etc)
charger_connector2_power_offered (or connector3 etc)
charger_connector2_time_session (or connector3 etc)
charger_connector2_transaction_id (or connector3 etc)
charger_connector2_current_import (or connector3 etc)
charger_connector2_current_offered (or connector3 etc)
charger_connector2_energy_session (or connector3 etc)
charger_connector2_status (or connector3 etc)
charger_connector2_stop_reason (or connector3 etc)
charger_connector2_error_code (or connector3 etc)
```

**Further thoughts**

It would be ideal if connector2, connector3 etc ... controls were only enumerated for chargers that support them.  Is this possible under the constraints of the OCPP integration?

Sensors for connector2, connector3 etc should not be provided as separate entities because the numbers begin to get overwhelming.  Rather, they should be extracted from extra_attr using Templates and displayed using a Lovelace Card Templater as described above.

Maybe we should abbreviate `connector` to `conn` so we end up with conn2, conn3 etc.

@lbbrhzn @drc38 looking forward to your comments on this first draft of the proposed naming convention.  If we can arrive at some sort of consensus, I could use it as the basis of a Wiki.

**Reply by drc38** [reply]
@GrantKeymer the naming looks sensible to me.  @lbbrhzn would extra connector switches be added in an options flow or dynamically somehow in post_connect?

**Reply by lbbrhzn** [reply]
I think the most user friendly way is to new entities dynamically, when a charger first connects. Maybe you just need a simple config screen to allow access and give the charger a name.

To avoid clutter we should only create entities that are needed and supported. So we need to read the feature profile and the number of connectors.

To avoid clutter we can store metric information in extra_attr, but we need to decide what metrics are stored in what entity. For instance, you could store all connector related metrics in the connector_status entity. @GrantKeymer, can you indicate which entity should be used for each metric? 

A disadvantage of using the extra_attr approach is that features like automatic unit conversion are not supported, and that a user will need to use a template or custom card to make the metric visible in the user interface. An alternative approach (which i've seen working quite well in some other custom integrations) is to create separate entities for each metric by default, but have many of them them disabled by default.  The user can then decide which entities should be enabled, and can easily disable them again if they turn out not to be interesting after all. Enabling / disabling is done from the device overview. We don't necessarily need to choose between one approach or the other, a combination is also possible.

@GrantKeymer , I noticed you have reordered the naming of some metric, i.e. charger_connector_status instead of charger_status_connector. This is also more intuitive to me. To make it more consistent, I propose to use charger_connector1_status etc. even if there is only one connector.


Taking about naming: we currently use the OCPP approach for naming the metrics, i.e. Energy.Import.Register. The ocpp package and home assistant automatically translate this to a pythonic naming scheme: energy_import_register. Maybe we should make things more consistent, and always use the pythonic name instead of the ocpp name?

**Reply by lbbrhzn** [reply]
I think it would be nice to also collect metrics per tag, for those chargers that support RFID tags.

These could be metrics like <tag>_session_start_time, <tag>_session_stop_time, <tag>_session_duration, <tag>_power, <tag>_energy_session, <tag>_energy_total. So, again the question here: separate entities, single entity with extra_attr, or both?

We could create a switch entity that controls authorization of a particular tag: <tag>_authorization. The first time a new tag is detected authorization will be automatically denied, but you can accept subsequent tries by flipping the switch. I like that better than manually maintaining an authorization list in an options flow. Home assistant already allows the user to give an entity a more meaningful name, so it should not be need to add additional functionality for that.

**Reply by GrantKeymer** [reply]
Thanks for your very detailed reply.

> I think the most user friendly way is to new entities dynamically, when a charger first connects. Maybe you just need a simple config screen to allow access and give the charger a name.

You've just reminded me of something I noticed recently...

When you add the OCPP integration, you have the option to change the name of all associated entities from the default.  Originally I chose `Wallbox` so all my entities were prefixed accordingly.  More entities were added later but these appeared with a prefix of `charger` instead of `Wallbox`.

I couldn't find any way to go back and rename the new ones with a prefix of `Wallbox` so I ended up removing the OCPP integration and adding it back again, this time leaving the name set to the default of `charger` so I don't have the same problem in future.

If there is any easy way to rename entities which are added later, I should document this in the Wiki, otherwise I should warn users about the difficulty they will experience if they change the device name from the default.

@lbbrhzn which do you think is the best approach here?

> To avoid clutter we should only create entities that are needed and supported. So we need to read the feature profile and the number of connectors.
> 
> To avoid clutter we can store metric information in extra_attr, but we need to decide what metrics are stored in what entity. For instance, you could store all connector related metrics in the connector_status entity. @GrantKeymer, can you indicate which entity should be used for each metric?

This sounds like a very good idea.  We are at the point now where many sensors have been added in an 'ad hoc' manner so some additional structure would be helpful.

@lbbrhzn - Where can I find the list of existing entities to put alongside each metric?  Or is this something we are looking to create now?

> 
> A disadvantage of using the extra_attr approach is that features like automatic unit conversion are not supported, and that a user will need to use a template or custom card to make the metric visible in the user interface. An alternative approach (which i've seen working quite well in some other custom integrations) is to create separate entities for each metric by default, but have many of them them disabled by default. The user can then decide which entities should be enabled, and can easily disable them again if they turn out not to be interesting after all. Enabling / disabling is done from the device overview. We don't necessarily need to choose between one approach or the other, a combination is also possible.

I agree that the need to use a template or custom card is a disadvantage and wasn't aware of the automatic unit conversion issue.  Perhaps it is best to do as you suggest, at least initially.  Keep things tidy by only having the most important entities enabled by default and leaving secondary ones disabled.  Users can then add a few extra entities, see if they find them useful and disable them again if not.
> 
> @GrantKeymer , I noticed you have reordered the naming of some metric, i.e. charger_connector_status instead of charger_status_connector. This is also more intuitive to me. To make it more consistent, I propose to use charger_connector1_status etc. even if there is only one connector.

Thanks!  I thought it was more logical too.  Good idea to change to connector1 for consistency.
> 
> Taking about naming: we currently use the OCPP approach for naming the metrics, i.e. Energy.Import.Register. The ocpp package and home assistant automatically translate this to a pythonic naming scheme: energy_import_register. Maybe we should make things more consistent, and always use the pythonic name instead of the ocpp name?

I've found the existing naming scheme confusing too.  I think we should adopt the pythonic naming approach because that is what users will be dealing with when they create their cards.  The underlying OCPP definitions are always there for reference if anyone is so inclined, but most users won't need to be aware of them.

**Reply by GrantKeymer** [reply]
> I think it would be nice to also collect metrics per tag, for those chargers that support RFID tags.
> 
> These could be metrics like _session_start_time, _session_stop_time, _session_duration, _power, _energy_session, _energy_total. So, again the question here: separate entities, single entity with extra_attr, or both?

I can see as time goes on, we will have more and more metrics to deal with.  However, as each new user comes along they would need to deal with templates to extract the affected metrics from extra_attr and then there is the issue of formatting for display, etc, etc.  Sounds like a recipe for a lot of problems.  However, it creates extra work for you devs if we decide to create a new entity each time a metric is added.

I think the most important thing will be to do what you suggested in your previous reply...

`...the most user friendly way is to new entities dynamically, when a charger first connects...`

This will keep the clutter from getting unmanageable.  Then adopt a case-by-case approach to whether a new entity is created for a metric or not.  If you think it's something most users would never need, just leave it stored in extra_attr.  Otherwise create a new entity.  Either way it needs to be documented in the Wiki, so please tag me in the Pull Request when you add a new metric.

@lbbrhzn - Where can I find a list of what is currently stored in extra_attr?  This is something I could put into the Wiki as time permits.
> 
> We could create a switch entity that controls authorization of a particular tag: _authorization. The first time a new tag is detected authorization will be automatically denied, but you can accept subsequent tries by flipping the switch. I like that better than manually maintaining an authorization list in an options flow. Home assistant already allows the user to give an entity a more meaningful name, so it should not be need to add additional functionality for that.

This sounds like a good way of managing the authorization list.  Much preferable to editing an options flow.

**Reply by lbbrhzn** [reply]
@GrantKeymer, 

The list of status we currently support can be found in the HAChargerStatuses class, which is defined in the enums.py file:
https://github.com/lbbrhzn/ocpp/blob/main/custom_components/ocpp/enums.py

The optional measurands we currently support are defined in MEASURANDS list in const.py:
https://github.com/lbbrhzn/ocpp/blob/main/custom_components/ocpp/const.py

The status sensors are always created, but the measurands are optional, and can be selected in the config flow.

The meter values statuses and error codes that are reported by the charger are stored internally in a table of metrics. A Metric simply has a name, value, unit and extra_attributes. Currently there is a 1 to 1 mapping from a metric to a sensor entity, which means adding an additional metric is not much work at all. I've got plans to extend the Metric class with a subscription / notification mechanism, so the sensors can be updated automatically when the properties of a metrics are updated. Currently we use brute force to update of all entities when certain messages are received.

You can find out what information is stored as extra attributes by searching for '.extra_attr' in api.py:
https://github.com/lbbrhzn/ocpp/blob/main/custom_components/ocpp/api.py

I haven't made a overview, but it would be very good to do!

**Reply by GrantKeymer** [reply]
> To avoid clutter we can store metric information in extra_attr, but we need to decide what metrics are stored in what entity. For instance, you could store all connector related metrics in the connector_status entity. @GrantKeymer, can you indicate which entity should be used for each metric?

After looking through the various lists and measurands, here's my suggestion.  Hopefully I've understood correctly but wouldn't be surprised if I didn't.

**Sensors relating to the entire charger
Stored in `charger_status`**

```
charger_frequency
charger_voltage
charger_energy_import_register
charger_energy_meter_start
charger_connectors
charger_status
charger_latency_ping
charger_latency_pong
charger_heartbeat
charger_reconnects
charger_id
charger_id_tag
charger_error_code
charger_model
charger_features
charger_serial
charger_status_firmware
charger_timestamp_config_response
charger_timestamp_data_response
charger_vendor
charger_version_firmware
```
**Sensors relating to a particular connector
Stored in `connector1_status` (or connector2_status etc)**

```
connector1_power_active_import
connector1_power_offered
connector1_time_session
connector1_transaction_id
connector1_current_import
connector1_current_offered
connector1_energy_session
connector1_status
connector1_stop_reason
connector1_error_code
```

How does that look?  Am I on the right track now?

**Reply by drc38** [reply]
> We could create a switch entity that controls authorization of a particular tag: _authorization. The first time a new tag is detected authorization will be automatically denied, but you can accept subsequent tries by flipping the switch. I like that better than manually maintaining an authorization list in an options flow. Home assistant already allows the user to give an entity a more meaningful name, so it should not be need to add additional functionality for that.

An actionable notification might work well for authorising dynamically:
https://companion.home-assistant.io/docs/notifications/actionable-notifications/

**Reply by GrantKeymer** [reply]
Here's what I've found so far regarding the usage of extra_attr...

`data_transfer` function stores vendor specific data from the charger in the form **{message id: data payload}**

`get_configuration` function stores **{key: value pairs}** for each config key supported by charger

`process_phases` function stores metering data from charger according to the following sequence:

**Name of metric** (e.g. Voltage)
**Phase** (L1, L2, L3)
**Value
Unit
Context** (Sampled Periodically or according to Clock)

`on_status_notification` function stores status changes relating to individual connectors in the form **{connector_id: status}**
Also stores error codes relating to individual connectors in the form **{connector_id: error_code}**

Other references to extra_attr seem to be retrieving the data rather than storing it.

**Reply by GrantKeymer** [reply]
At last, I think I've got my head around this now.  What we need is to reduce the number of entities and use `extra_attr` for some of them, to store diagnostic data which won't need to be separately displayed/charted in HA.  How about this...

**Sensors relating to the entire charger**

Stored in entity `charger_metrics`

```
voltage (top level variable i.e. chartable directly)
frequency
energy_import_register
energy_meter_start
status
```

Stored in entity `charger_diagnostics`

```
heartbeat (top level variable i.e. chartable directly)
error_code
latency_ping
latency_pong
reconnects
id_tag
timestamp_config_response
timestamp_data_response
```

Stored in entity `charger_info`

```
vendor (top level variable i.e. displayable directly)
model
id
serial
features
status_firmware
version_firmware
connectors
```

**Sensors relating to a particular connector**

Stored in `connector1_xxx` (or connector2_xxx etc) as separate entities

```
connector1_power_active_import
connector1_power_offered
connector1_current_import
connector1_current_offered
connector1_energy_session
connector1_time_session
```
Stored in `connector1_status` (or connector2_status etc)

```
status (top level variable i.e. displayable directly)
transaction_id
stop_reason
error_code
```

I've left the first 6 connector-specific entities separate for now because there is often the need to use them for a meter, chart, status indicator etc.  If hidden inside extra_attr, it becomes more difficult to do this.  However, the last few (status, transaction_id, stop reason and error_code) can sensibly be combined.

Hopefully this will be more what you had in mind.

**Reply by drc38** [reply]
If we will need to create a template sensor (which is sensor entity) to chart data I assume there is no particular benefit in storing it in `extra_attr` or is there?

The charger info data makes sense combing into a single metric. Some of the info however already is available in the device registry such as firmware, so do we need this in the metric?

**Reply by GrantKeymer** [reply]
> If we will need to create a template sensor (which is sensor entity) to chart data I assume there is no particular benefit in storing it in `extra_attr` or is there?

I don't think we need to create a template sensor to chart each of these less important variables.  It would be good to provide one as an example, which could be documented in the Wiki.  The benefit I see from storing them in `extra_attr` is that it's very easy to use the dropdown `attributes` arrow to display the extra variables associated with a given entity.  If someone decides they need to chart any particular `extra_attr` variable, they can create their own template, modelled on the example.
> 
> The charger info data makes sense combing into a single metric. Some of the info however already is available in the device registry such as firmware, so do we need this in the metric?

Agreed, we could leave pretty much all charger info contained solely in the registry.

**Comment by GrantKeymer** [comment]
OK guys, I'm ready to start writing the Wiki whenever you give me the green light.  Are we agreed on the above list of sensors + entities as being the best way forward?  If not, then let's refine it a bit more until we reach agreement.

After looking for suitable examples of Wikis hosted on Github, I've chosen this one as a good starting point (just the structure not any of the content)...

**https://github.com/Netflix/Hystrix/wiki**

One of the first sections I want to write, deals with pitfalls best avoided during installation.  Basically, things I wish I'd known before installing this integration for the first time.  To help me along the way, I need your comments on this issue...

When you add the OCPP integration, you have the option to change the name of all associated entities from the default. Originally I chose Wallbox so all my entities were prefixed accordingly. More entities were added later but these appeared with a prefix of charger instead of Wallbox.

I couldn't find any way to go back and rename the new ones with a prefix of Wallbox so I ended up removing the OCPP integration and adding it back again, this time leaving the name set to the default of charger so I don't have the same problem in future.

If there is any easy way to rename entities which are added later, I should document this in the Wiki, otherwise I should warn users about the difficulty they will experience if they change the device name from the default.

@lbbrhzn @drc38 - **Let me have your comments please**

**Reply by drc38** [reply]
There's no easy way to update sensor names that I'm aware of. As long as you use a consistent name for the charger every time you reinstall you're ok. The integration can support multiple chargers by giving them unique names.

**Reply by GrantKeymer** [reply]
My point is that if you want to avoid re-installing, rather just taking the integration updates via HACS as they become available, then you don't have any control over the prefix used for newly added entities.  They all end up with a prefix of `charger.`

Therefore, it's probably best to leave the device name `charger` in place so all entities have the same prefix.  Does that make sense?

**Reply by drc38** [reply]
> My point is that if you want to avoid re-installing, rather just taking the integration updates via HACS as they become available, then you don't have any control over the prefix used for newly added entities. They all end up with a prefix of `charger.`
> 
> Therefore, it's probably best to leave the domain name `charger` in place so all entities have the same prefix. Does that make sense?

They shouldn't,  I've never had that issue and always used a non-default charger name.

If it can happen though worth mentioning.

**Reply by GrantKeymer** [reply]
Hmmm, that's interesting.  It definitely happened to me with the recently added _status_connector and _error_connector entities.

**Comment by lbbrhzn** [comment]
Please note there are two places where a name/identity for a charger can be entered. The charger name in the configuration panel and the charger identity (entered when configuring the url). We currently use the config name for creating the entities, but use the charger identity for the connection. There is code in place for supporting multiple chargers, but that has not been tested and I don't think it works.

**Reply by drc38** [reply]
I thought @mcfranky had more than 1 connected?

**Reply by GrantKeymer** [reply]
> Please note there are two places where a name/identity for a charger can be entered. The charger name in the configuration panel and the charger identity (entered when configuring the url). We currently use the config name for creating the entities, but use the charger identity for the connection.

Excellent, thanks for clarifying that.

**Reply by McFranky** [reply]
@drc38 I have 7 instances of Server/Client installed.
![grafik](https://user-images.githubusercontent.com/59682922/153796808-2ac0212f-cd0a-4797-8130-9f9587a9d2ff.png)

**Comment by GrantKeymer** [comment]
@lbbrhzn - When you're ready, please create a blank Wiki page in .MD format with your OCPP logo at the top.  I'll fork and start work from there.  Thanks!

**Comment by GrantKeymer** [comment]
@lbbrhzn - Thanks for creating the blank Wiki page with logo for me.  Now I have my first page ready to be merged with your repo but I can't seem to make a Pull Request to action this.  It keeps saying that...

```
There isn’t anything to compare.
lbbrhzn:main is up to date with all commits from GrantKeymer:main
```

Probably me not understanding how PRs work, but I just wonder, do they only work with code and not documentation?

Here is my Wiki page... https://github.com/GrantKeymer/ocpp/wiki

**Comment by lbbrhzn** [comment]
Many thanks for the updated documentation, great job!
Unfortunately, looks like PRs only only work for the main repo, not for the wiki. There are some workarounds in use, see:
https://stackoverflow.com/questions/10642928/how-to-pull-request-a-wiki-page-on-github
Not sure what would be the best approach in our case.

**Reply by lbbrhzn** [reply]
Maybe this approach would work for us:

https://nimblehq.co/blog/create-github-wiki-pull-request

**Reply by lbbrhzn** [reply]
Another option would be to use readthedocs: https://docs.readthedocs.io/en/stable/tutorial/
In this case the documentation will be generated from .md and or .rst files in the /docs folder, and published as static html pages on readthedocs.io. It also generates pdf and epub output. It is based on sphynx, like the github wiki system.

This definitely looks more powerful that using the built-in wiki, but perhaps overkill for this project.

**Reply by GrantKeymer** [reply]
> Maybe this approach would work for us:
> 
> https://nimblehq.co/blog/create-github-wiki-pull-request

I like this one best.  It's quite simple and works through PRs in the normal way, rather than needing to create an 'Issue' whenever I have an update.  We don't want to clutter up the Issues page with trivial requests that aren't really issues at all.

Well done on finding 3 different options 👍

**Reply by lbbrhzn** [reply]
I think we should be able to do both: push to github wiki, and publish on readthedocs. 
I've setup a readthedocs project, and did an intial build. See https://home-assistant-ocpp.readthedocs.io/en/latest/#

**Reply by GrantKeymer** [reply]
Fair enough, it looks good.  I was just concerned that having it in two places might create double the amount of work, but if you're happy on that score, then all good.

Each approach has its advantages...

1)  It's nice to utilise the built-in Wiki page because everything is kept under the GitHub umbrella which is very efficient on low-end devices.

2)  It's also nice to have the ability to export the Docs as PDF, which ReadTheDocs provides.

So I guess this way, we have the best of both worlds so long as you're happy doing two lots of pushes.

**Reply by lbbrhzn** [reply]
Publishing to readthedocs has some more benefits: it's searchable, and will be indexed by search engines. The build-and-publish step is done automatically when a pull request is merged to master, so this does not require additional work.
The push to the github wiki can also be done automatically, i'll look into that next.

**Reply by GrantKeymer** [reply]
Great to see the [Wiki](https://github.com/lbbrhzn/ocpp/wiki/user-guide) now has more functionality, including the User Guide page I wrote.  Near the bottom is a section entitled `Useful Entities for Wallbox Pulsar Plus`.  It would be great to see equivalent lists for other supported chargers.

@drc38 - Wondering if you'd like to compile a list of useful entities for EVNEX?
@dreimer1986 - Wondering if you'd like to compile a list of useful entities for EVBox Elvi?

Perhaps just post here for now until Pull Requests are working for the Wiki.

**Reply by dreimer1986** [reply]
@GrantKeymer I can of course tell you what entites are doing something here, but I have the Elvi without any digital three-phase meter. Thus mine will show less than one with the meter inside. Telling you which might be useful... I am a Information freak, thus I always log everything. ^^ My favorite is of course number.evbox_maximum_current. It made me sit down and create some fuse protection. 

```
sensor:
  - platform: template
    sensors:
      maxcharge:
        friendly_name: "Max Laden"
        unit_of_measurement: "A"
        value_template: >
          {% if (15000 - states('sensor.powerhaus_power')|float) / 3 / 230 > 16.0 %}
            16.0
          {% else %}
            {{ (((15000 - states('sensor.powerhaus_power')|float) / 3 / 230)|round(0)*10)/10 }}
          {% endif %}
```

```
- id: schutz_vor_ueberlast
  alias: Schutz vor Überlast
  trigger:
  - platform: state
    entity_id: binary_sensor.max_laden_ok
    to: 'off'
  condition:
  - condition: state
    entity_id: sensor.evbox_status_connector
    state: 'Charging'
  action:
  - service: number.set_value
    data:
      entity_id: number.evbox_maximum_current
      value: '{{ states("sensor.maxcharge") }}'
  - service: notify.alexa_media
    data:
      target:
      - media_player.daniels_echo_dot
      - media_player.daniels_echo_show
      data:
        type: announce
      message: Maximaler Ladestrom wurde geändert auf {{ states('sensor.maxcharge')|int }} Ampere!
  - service: notify.telegrambot
    data:
      message: Maximaler Ladestrom wurde geändert auf {{ states('sensor.maxcharge')|int }} Ampere!
  mode: single
```

SO enough of my not asked for stuff... these work:

```
switch.evbox_availability
Works fine. Off when something causes a problem or a reboot etc.

switch.evbox_charge_control
Works fine. On when charging

sensor.evbox_connectors
Always 1 here.

sensor.evbox_current_offered
Seems to work. Shows normally 16 A here when charging.

sensor.evbox_error_code
Normally NoError, so works.

sensor.evbox_error_code_connector
Normally NoError, so works.

sensor.evbox_features
Always shows Profiles.AUTH|REM|RES|SMART|FW|CORE, so works.

sensor.evbox_heartbeat
Works fine. Timestamp is fine.

sensor.evbox_id
Always shows EVB-PXXXXXXXX aka my serial number, works fine.

sensor.evbox_id_tag
Shows the HEX ID of the key which starts the charging. I have coded one inside to make it work without any need for a key. Works fine.

sensor.evbox_latency_ping
Works fine.

sensor.evbox_latency_pong
Works fine.

number.evbox_maximum_current
FINALLY WORKS FINE!!! :D

sensor.evbox_model
Always shows G4E-WIFI, seems to be the internal codename? Works fine.

sensor.evbox_reconnects
Works fine. Counter is sometimes low, sometimes not. Bad WIFI == weather is a problem ^^

button.evbox_reset
Works fine. Does a reset ^^

sensor.evbox_serial
Same as the ID, but without the EVB-P, so the real serial. Works fine.

sensor.evbox_status
Normally shows Available, works fine.

sensor.evbox_status_connector
Available, Charging, SuspendedEV, all fine.

sensor.evbox_status_firmware
Normally unknown, but was informative when I did the firmware update. Works fine.

sensor.evbox_stop_reason
Seems to work fine. Shows EVDisconnected when I expect it to.

sensor.evbox_temperature
Works fine when charging.

sensor.evbox_time_session
Shows the time from car plugged in until disconnected, works fine.

sensor.evbox_timestamp_config_response
Works fine. Timestamp is fine.

sensor.evbox_timestamp_data_transfer
Works fine. Timestamp is fine.

sensor.evbox_transaction_id
Some other ID when charging?

button.evbox_unlock
Never had a need to... Likely works.

sensor.evbox_vendor
Always shows EV-BOX ^^

sensor.evbox_version_firmware
Shows the current firmware version. Works great.
```

Does not work here. Most of them likely work fine on a box with three-phase-meter:

```
sensor.evbox_current_import
sensor.evbox_energy_active_import_register
sensor.evbox_energy_meter_start
sensor.evbox_energy_session
sensor.evbox_frequency
sensor.evbox_power_active_import
sensor.evbox_timestamp_data_response
sensor.evbox_voltage
```

**Reply by drc38** [reply]
@GrantKeymer unfortunately there is not much point documenting the Evnex as they no longer support using a custom Ocpp backend for new customers.

**Reply by GrantKeymer** [reply]
> @GrantKeymer unfortunately there is not much point documenting the Evnex as they no longer support using a custom Ocpp backend for new customers.

Oh, sorry to hear that.  Seems a very backward step for such an innovative company to make.  Locking things and keeping them proprietary is not helpful for tinkerers like us.  Glad to see others are sticking with OCPP but sad that NZ's only EV charger manufacturer is not among them.  I guess you're OK as you got your EVNEX before they made this decision?

**Reply by drc38** [reply]
As long as I don't update the firmware or need to change my HA ip address then I'll be ok continuing to use the Evnex. I think the lock-in approach will hurt them in the long run as chargers are just one part of the future demand response eco-system.

**Reply by lbbrhzn** [reply]
@GrantKeymer the automatic push to wiki is now working as well.


The whole procedure for updating the documentation works similar to updating the source code: 

- fork the repo (if not done already)
- clone the fork (if you want to work locally)
- create a new branch for storing your changes 
- change the .md files in the /docs folder on the branch (either locally or through the GitHub web interface)
- add, commit and push your changes (local)
- create a pull request

The pull request can then be reviewed, and merged to master after approval. The merge will trigger an update to readthedocs and to the wiki of lbbrhzn/ocpp.
When a new release is made, the documentation corresponding to that release will be added to readthedocs.
The wiki will always reflect the latest changes.
Note that updating the wiki directly (instead of using /docs) will not work. The /docs folder is leading and will overwrite whatever is in the wiki.

**Reply by GrantKeymer** [reply]
> @GrantKeymer the automatic push to wiki is now working as well.
> 
> The whole procedure for updating the documentation works similar to updating the source code:
... 
Note that updating the wiki directly (instead of using /docs) will not work. The /docs folder is leading and will overwrite whatever is in the wiki.

Excellent!  Thanks for the very detailed explanation.  I'll give it a try in the next few days.

**Reply by GrantKeymer** [reply]
> My favorite is of course number.evbox_maximum_current. It made me sit down and create some fuse protection.
>
Haha, I love the overcurrent warning using Alexa and Telegram :rofl:
 
Thanks for all the other explanations.  That will make for a very detailed Wiki page 👍

**Reply by GrantKeymer** [reply]
> @GrantKeymer the automatic push to wiki is now working as well.
> 
> The whole procedure for updating the documentation works similar to updating the source code:
> 
> * fork the repo (if not done already)
> * clone the fork (if you want to work locally)
> * create a new branch for storing your changes
> * change the .md files in the /docs folder on the branch (either locally or through the GitHub web interface)
> * add, commit and push your changes (local)
> * create a pull request
> 
> The pull request can then be reviewed, and merged to master after approval. The merge will trigger an update to readthedocs and to the wiki of lbbrhzn/ocpp. When a new release is made, the documentation corresponding to that release will be added to readthedocs. The wiki will always reflect the latest changes. Note that updating the wiki directly (instead of using /docs) will not work. The /docs folder is leading and will overwrite whatever is in the wiki.

@lbbrhzn - I'm finally coming back to this after getting other stuff out of the way.  I've deleted the old repo fork I made some time ago because it was well out of date, then created a fresh fork.  I don't need to work locally so I didn't clone the fork.  However, I can't see how to `Create a new branch` even after looking at the GitHub Docs.  This is the nearest option I can find...

![GitHub - Rename Branch](https://user-images.githubusercontent.com/35156198/160232921-c8141fd3-c353-4bb6-8749-d8941adf268f.png)

I suspect it's the wrong option to use, but can't see any way to `Create a new branch` as instructed.  Thanks 🤦‍♂️

**Comment by lbbrhzn** [comment]
hi @GrantKeymer , please note that you should not fork the wiki, but the code repository.
You can then create a new branch by clicking on 'main'. Then enter your branch name, and select 'create branch'.

![afbeelding](https://user-images.githubusercontent.com/8673442/160238182-56137538-0b8d-4595-9a9c-dcafdc4dcc42.png)
